### DISCUSSION TITLE
Replay 2.0

### DISCUSSION THREAD

**Discussion by Cookiesmuch** [ppy/osu#21729] [Ideas]
_Hello! My name is Cookiesmuch, and I'm a regular Lazer! Player, I would like to recommend a few possible additions to the replay system that could improve its functionality and overall Quality Of Life._

#### To start off, the Replay System is quite mature (since it already exists within the Older (Stable) branch of Osu!
This, in no way shape or form, is a refactor or a replacement for that system.
I would like to disclaim, however, that it (in my biased opinion) would be for the benefit of all concurrent and future Lazer! players.

## Small Additions

- ### Settings Loggers
  - #### Skin Logger/Skin Saver
    - The Skin Logger/Saver would do as it says; Logs or Saves the Skin that the player had been using when the play had been made.
    - There could be an option to automatically swap to the Skin when viewing the Replay.
    - Im fairly certain that this could easily be done by saving the skin's name and/or an assigned Skin ID in the Replay's savefile.

  - #### Other Settings Logger/Saver
    - Similarly (or in the first place, this could have been put in the same item as the previous one), This saves the other quick setting options such as Snaking Sliders, Color Normalization, Background Dim/Blur, Beatmap Storyboard/Skin/Color, Beatmap Hitsounds, and other quick settings.
    - Once again, could automatically swap to those options when viewing the replay.
    - Could probably also just be a few variables stored in the Replay file.

  - #### Quick Swapping
    - To Elaborate further on the swapping, there could be an option (or keybind) to automatically switch to the replay's logged settings.
    - On the Flip Side, there could be an option to automatically switch back to the user's settings before the replay's was set.

- ### Replay Sharing
  - Quickly adds the option to export the Replay File to be able to Upload/Share the Replay like how you could in stable.

- ### Replay/Gameplay Commenting.
  - Allow the User to add comments to timestamps, so they could come back to the timestamp in the future, or comment their mistakes so that they could come back and improve on their mistakes.
  - Comment Panel that allows the player to see all the comments they have on their account?

## Big Additions

- ### Replay Exporting (Not Like the Sharing)
  - #### Replay Snipping
    - Allow the User to Trim a Replay from Timestamp-to-Timestamp to effectively only have the segment they want. These Timestamps could also be at comments. (starts and or ends at a comment)
    - Allow the User to then save this as a separate replay.

  - #### Replay Export Panel
    - Adds a new Panel to the Replay that the User can pull up, In this panel, The User could fine tune the Settings Used (Such as those mentioned in the Settings Logger), and Video Export Settings.
    - Video Export Settings would include The Resolution, File Type (or It could be limited to MP4 to lighten the implementation), Frame Rate, Bitrate, and others.
    - **Video Encoder** Finally, what has been awaited for this entire post, The Video Encoder would Asynchronously Export the Replay at the Target Resolution And Frame Rate. This could also been done synchronously by playing the Replay and “Recording it in real time”, however the Asynchronous Option would allow it to export in Higher Resolution and Frame Rate than the PC could handle; It would simply just take longer to Export, Just like Video Editing Software.

# In Conclusion, this is a Compilation Post for possible Improvements on the Replay System, That could benefit Normal Players, Pro Players, and Content Creators Alike. Allowing people to look back and improve on their mistakes, or get footage of their Gameplay, without sacrificing some performance while they play.

If anyone has any suggestions, additions, or objections to this post, kindly just comment, and I could add it to the post.

**Comment by peppy** [comment]
All of this sounds interesting (and most have been suggested in the past). To give you an idea of direction, we already have intentions on adding skin details (and make the game download the skin when watching). Things like commenting, which is in stable but not really used and likely going to be deprecated.

**Comment by ghost** [comment]
By the way, I think it's worth mentionning you can already export replays from lazer. Simply go to the song select screen, find the map which has the replay you're interested in exporting, find that score in the leaderboard, then right-click and then click "Export" under the menu.

**Reply by Cookiesmuch** [reply]
the export i had intended would be an actual video export, such as exporting to mp4 and etc.

**Reply by Theighlin** [reply]
This has been asked multiple times and the current stance is that it's out of scope of the game and won't be added.

**Comment by Cookiesmuch** [comment]
https://github.com/nahkd123/osu-replay-viewer hmmmmmmm

**Reply by Cookiesmuch** [reply]
Gonna mark this as an answer.

**Reply by peppy** [reply]
Yeah I don't think that's an answer. Let's leave this thread open until we actually realise some of the features in the game.
